stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -2535,10 +3132,92 @@ impl NakamotoChainState {
+                &new_tip.consensus_hash,
+            ));
+            marf_values.push(nakamoto_keys::make_block_id_value(&new_tip.block_id()));
+
+            // record coinbase height of this tenure
+            marf_keys.push(nakamoto_keys::coinbase_height(&new_tip.consensus_hash));
+            marf_values.push(nakamoto_keys::make_u64_value(coinbase_height));
+
+            // record that this previous tenure is done
+            let Some(tenure_change_tx) = new_block.get_tenure_change_tx_payload() else {
+                // should be unreachable
+                error!(
+                    "Block {} is a tenure-change block, but does not contain a tenure-change tx",
+                    new_tip.block_id()
+                );
+                return Err(ChainstateError::InvalidStacksBlock(
+                    "Tenure-change block does not have a tenure-change tx".into(),
+                ));
+            };
+
+            marf_keys.push(nakamoto_keys::finished_tenure_consensus_hash(
+                &tenure_change_tx.prev_tenure_consensus_hash,
+            ));
+            marf_values.push(nakamoto_keys::make_bool_value(true));
+
+            // record parent tenure linkage
+            marf_keys.push(nakamoto_keys::parent_tenure_consensus_hash(
+                &tenure_change_tx.tenure_consensus_hash,
+            ));
+            marf_values.push(nakamoto_keys::make_consensus_hash_value(
+                &tenure_change_tx.prev_tenure_consensus_hash,
+            ));
+
+            // record last block-found tenure
+            let block_found_tenure_id = NakamotoTenureId {
+                burn_view_consensus_hash: tenure_change_tx.burn_view_consensus_hash.clone(),
+                block_id: new_tip.block_id(),
+            };
+
+            marf_keys.push(nakamoto_keys::block_found_tenure_id(
+                &tenure_change_tx.tenure_consensus_hash,
+            ));
+            marf_values.push(nakamoto_keys::make_tenure_id_value(&block_found_tenure_id));
+        }
+
+        if let Some(tenure_tx) = new_block.get_tenure_tx_payload() {
+            // either a block-found or a tenure-extend, but we have a new tenure ID in this fork
+            let tenure_id = NakamotoTenureId {
+                burn_view_consensus_hash: tenure_tx.burn_view_consensus_hash.clone(),
+                block_id: new_tip.block_id(),
+            };
+
+            marf_keys.push(nakamoto_keys::ongoing_tenure_id().to_string());
+            marf_values.push(nakamoto_keys::make_tenure_id_value(&tenure_id));
+        }
+
+        // record the highest block in this tenure
+        marf_keys.push(nakamoto_keys::highest_block_in_tenure(
+            &new_block.header.consensus_hash,
+        ));
+        marf_values.push(nakamoto_keys::make_block_id_value(&new_tip.block_id()));
+
+        for (key, value) in marf_keys.iter().zip(marf_values.iter()) {
+            debug!("Set Nakamoto MARF pair '{}' = '{}'", &key, &value);
+        }
```

Comment:
This should be condensed into a single log invocation, otherwise it will add a pretty large volume of debug logs.

```suggestion
            debug!("Set Nakamoto headers MARF"; "keys" => ?marf_keys, "values" => ?marf_values);
```

Reply:
Didn't know you could do that in a K/V log

Reply:
Well, that syntax is really just printing the `debug` fmt -- prefixing the values with `?` give you debug strings, and `%` give you fmt strings.